model_compression_toolkit/data_generation/keras/optimization_utils.py

```diff
@@ -0,0 +1,436 @@
+        self.n_batches = len(self.batch_opt_holders_list)
+        self.random_batch_reorder()
+        self.all_imgs_stats_holder = KerasAllImagesStatsHolder(n_batches=self.n_batches,
+                                                               batch_size=self.batch_size,
+                                                               mean_axis=self.mean_axis)
+        if self.use_all_data_stats:
+            for i_batch in range(self.n_batches):
+                input_imgs = self.image_pipeline.image_output_finalize(images=self.get_images_by_batch_index(i_batch))
+                output = activation_extractor.run_model(inputs=input_imgs)
+                batch_stats_holder = self.all_imgs_stats_holder.batches_stats_holder_list[i_batch]
+                self.all_imgs_stats_holder.update_batch_stats(batch_stats_holder=batch_stats_holder,
+                                                              input_imgs=input_imgs,
+                                                              activation_extractor=activation_extractor)
+
+    def clip_and_reflect(self,
+                         z: tf.Tensor) -> tf.Tensor:
+        """
+        Clips and optionally reflects the input tensor `z` channel-wise based on the valid value range.
+
+        Args:
+            z (tf.Tensor): Input tensor to be clipped and reflected.
+
+        Returns:
+            tf.Tensor: Clipped and reflected tensor.
+        """
+        if self.clip_images:
+            images = z.numpy()
+            for i_ch in range(len(self.valid_grid)):
+                # Clip the values of the channel within the valid range.
+                clamp = tf.clip_by_value(t=z[:, :, :, i_ch], clip_value_min=self.valid_grid[i_ch][0],
+                                         clip_value_max=self.valid_grid[i_ch][1])
+                if self.reflection:
+                    # Reflect the values.
+                    images[:, :, :, i_ch] = 2 * clamp - z[:, :, :, i_ch]
+                else:
+                    images[:, :, :, i_ch] = clamp
+            # Assign the clipped reflected values back to `z`.
+            z.assign(images)
+            return z
+        else:
+            return z
+
+    def get_accumulated_stats_per_layer(self, layer_name: str) -> Tuple[tf.Tensor, tf.Tensor]:
+        """
+        Get the accumulated activation statistics for a layer.
+
+        Args:
+            layer_name (str): the name of the layer.
+
+        Returns: Tuple[tf.Tensor, tf.Tensor]: The averaged activation statistics (mean and second moment) on all the
+        batches for the specified layer.
+        """
+        total_mean, total_second_moment = 0, 0
+        for i_batch in range(self.n_batches):
+            batch_stats_holder = self.all_imgs_stats_holder.batches_stats_holder_list[i_batch]
+            mean, second_moment = self.all_imgs_stats_holder.get_stats(batch_stats_holder=batch_stats_holder,
+                                                                       layer_name=layer_name)
```

Comment:
I don't understand this code. all_images_stats_holder should have that accumulated stats across all batches. You shouldn't do the calculation here. This function is just a "get" function not a "compute" function.